base/not_intersecting_intervals.hpp

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "base/assert.hpp"
+
+#include <set>
+
+namespace base
+{
+/// \brief Implements data structure, that allows to add intervals with |AddInterval()|
+///        method. This method returns true if all previous intervals, which were added
```

Comment:
s/if/iff/
What do you mean by "which were added"? Does AddInterval always add an interval?

Reply:
More importantly, this looks like method documentation.
For the data structure documentation, it is enough to say
```
// [...] data structure that maintains a set of non-intersecting intervals.
// A new interval may only be added to the set if it does not intersect any
// of the present intervals, thus the choice which intervals to keep is made
// in a greedy online fashion with no lookahead.
```

Reply:
```cpp
namespace base
{
///\brief Data structure that maintains a set of non-intersecting intervals.
///       A new interval may only be added to the set if it does not intersect any
///       of the present intervals, thus the choice which intervals to keep is made
///       in a greedy online fashion with no lookahead.
template <typename T>
class NotIntersectingIntervals
{
public:
  /// \return true if there are no such intervals, that intersect the [left, right] interval.
  bool AddInterval(T left, T right);
```